factor/sms/classes/factor.php

```diff
@@ -0,0 +1,241 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * SMS Factor class.
+ *
+ * @package     factor_sms
+ * @subpackage  tool_mfa
+ * @author      Peter Burnett <peterburnett@catalyst-au.net>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace factor_sms;
+
+defined('MOODLE_INTERNAL') || die();
+
+use tool_mfa\local\factor\object_factor_base;
+
+class factor extends object_factor_base {
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function login_form_definition($mform) {
+
+        $mform->addElement('text', 'verificationcode', get_string('verificationcode', 'factor_sms'),
+            ['autocomplete' => 'one-time-code']);
+        $mform->setType("verificationcode", PARAM_ALPHANUM);
+        return $mform;
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function login_form_definition_after_data($mform) {
+        $this->generate_and_sms_code();
+        return $mform;
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function login_form_validation($data) {
+        $return = array();
+
+        if (!$this->check_verification_code($data['verificationcode'])) {
+            $return['verificationcode'] = get_string('error:wrongverification', 'factor_email');
+        }
+
+        return $return;
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function get_all_user_factors($user) {
+        global $DB;
+
+        $records = $DB->get_records('tool_mfa', array(
+            'userid' => $user->id,
+            'factor' => $this->name,
+        ));
+
+        if (!empty($records)) {
+            return $records;
+        }
+
+        // Null records returned, build new record.
+        $record = array(
+            'userid' => $user->id,
+            'factor' => $this->name,
+            'createdfromip' => $user->lastip,
+            'timecreated' => time(),
+            'revoked' => 0,
+        );
+        $record['id'] = $DB->insert_record('tool_mfa', $record, true);
+        return [(object) $record];
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function has_input() {
+        return true;
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function has_setup() {
+        return true;
```

Comment:
This looks quite promising:

https://aws.amazon.com/blogs/messaging-and-targeting/improve-sms-message-delivery-with-phone-number-verify/

> With Phone Number Verify, Amazon Pinpoint checks to see whether the phone number provided is in a valid format based on its country code prefix. If the number is formatted correctly, Amazon Pinpoint leaves the number as the customer entered it. If the number isn’t formatted correctly, Amazon Pinpoint compares the phone number to various rules to make it valid. For example, if a customer enters a leading zero after the country code, Phone Number Verify removes the extra character to make the number valid. In this case, if the customer entered +1 0 206 555 0199 (a United States phone number in the 206 area code), Phone Number Verify changes the number to a correctly formatted number without the extra zero (+1 206 555 0199).

Reply:
I'll make this a new issue for later

Reply:
https://github.com/catalyst/moodle-tool_mfa/issues/227